fastapi/pools/fastapi_app.py

```diff
@@ -0,0 +1,136 @@
+    """
+
+    def __init__(self):
+        self._queue_by_db_by_root_path: dict[str, dict[str, queue.Queue[FastAPI]]] = (
+            defaultdict(lambda: defaultdict(queue.Queue))
+        )
+        self.__cache_sequences = {}
+        self._lock = threading.Lock()
+
+    def __get_pool(self, env: Environment, root_path: str) -> queue.Queue[FastAPI]:
+        db_name = env.cr.dbname
+        return self._queue_by_db_by_root_path[db_name][root_path]
+
+    def __get_app(self, env: Environment, root_path: str) -> FastAPI:
+        pool = self.__get_pool(env, root_path)
+        try:
+            return pool.get_nowait()
+        except queue.Empty:
+            env["fastapi.endpoint"].sudo()
+            return env["fastapi.endpoint"].sudo().get_app(root_path)
+
+    def __return_app(self, env: Environment, app: FastAPI, root_path: str) -> None:
+        pool = self.__get_pool(env, root_path)
+        pool.put(app)
+
+    @contextmanager
+    def get_app(
+        self, env: Environment, root_path: str
+    ) -> Generator[FastAPI, None, None]:
+        """Return a  FastAPI app to be used in a context manager.
+
+        The app is retrieved from the pool if available, otherwise a new one is created.
+        The app is returned to the pool after the context manager exits.
+
+        When used into the FastApiDispatcher class this ensures that the app is reused
+        across multiple requests but only one request at a time uses an app.
+        """
+        self._check_cache(env)
+        app = self.__get_app(env, root_path)
+        try:
+            yield app
+        finally:
+            self.__return_app(env, app, root_path)
+
+    def get_cache_sequence(self, key: str) -> int:
+        with self._lock:
+            return self.__cache_sequences.get(key, 0)
+
+    def set_cache_sequence(self, key: str, value: int) -> None:
+        with self._lock:
+            if (
+                key not in self.__cache_sequences
+                or value != self.__cache_sequences[key]
+            ):
+                self.__cache_sequences[key] = value
+
+    def _check_cache(self, env: Environment) -> None:
+        cache_sequences = env.registry.cache_sequences
+        for key, value in cache_sequences.items():
+            if (
+                value != self.get_cache_sequence(key)
+                and self.get_cache_sequence(key) != 0
+            ):
+                _logger.info(
+                    "Cache registry updated, reset fastapi_app pool for the current "
+                    "database"
+                )
+                self.invalidate(env)
+            self.set_cache_sequence(key, value)
+
+    def invalidate(self, env: Environment, root_path: str | None = None) -> None:
+        db_name = env.cr.dbname
+        if root_path:
+            self._queue_by_db_by_root_path[db_name][root_path] = queue.Queue()
+        elif db_name in self._queue_by_db_by_root_path:
+            del self._queue_by_db_by_root_path[db_name]
```

Comment:
@lmignon Why do you think it is sufficient to delete content of _queue_by_db_by_root_path (ASGIMiddleware) this way? Threads created before supposed to be cleaned by Python GC? I've done some test with Odoo cache invalidation and logging current threads number, the number is incremented by 1  after every "Caches invalidated" signalling. The threads remain in the process.

Reply:
![Screenshot 2025-03-17 at 11 08 05](https://github.com/user-attachments/assets/020890f1-d0a3-463e-b35b-5f7893a83d49)

![Screenshot 2025-03-17 at 11 10 03](https://github.com/user-attachments/assets/f640ecf4-5507-4493-8f61-a0e783219daa)
the difference between 2 logs is 1 Odoo cache invalidation.

Reply:
@lmignon I think I've found the issue. seems like you should have beer redefine "__init__" in ASGIMiddleware
(https://github.com/abersheeran/a2wsgi/blob/master/a2wsgi/asgi.py#L130)
because every time new loop is created 

![Screenshot 2025-03-17 at 11 54 33](https://github.com/user-attachments/assets/f93a5d8e-8096-4f10-b65e-f115cf61f1bf)

Reply:
@veryberry The middleware used is defined [here](https://github.com/OCA/rest-framework/pull/499/files#diff-9681ad18c667fdc32a0df20eeb26ed638016005f61a5feef05266689f34d14c9R21) and extend https://github.com/abersheeran/a2wsgi/blob/master/a2wsgi/asgi.py#L130. The loop used comes from a poo and if one is available into the pool it will be reused.

Reply:
@lmignon the problem is that you have extended its __call__ method
But actually the "__init__" is triggered first, and there loop is alway None, and new loop is created every time you call  `return ASGIMiddleware(app)`. 
I've proved in locally controlling threads after Odoo cache invalidation.

Reply:
see https://github.com/OCA/rest-framework/pull/508/files

Reply:
Thanks
@lmignon just wanted you to be aware about this issue and know your opinion on it. In my case I move context manager usage from `__call__` to `__ini__`. It helped.

Reply:
Think of the event loop thread like a single train track. If two trains (calls) try to use the same track at the same time, they’ll collide, and everything will break. To keep things running smoothly, we let one train (call) complete its journey before allowing another one on the track.